contracts/lib.rs

```diff
@@ -469,258 +462,257 @@ pub mod prosopo {
+                lookup.unwrap()
+            };
+            if new {
+                self.provider_state_insert(&old_provider)?;
+            }
+            let mut new_provider = old_provider.clone();
+
+            // update the config
+            new_provider.service_origin =
+                service_origin.unwrap_or(old_provider.service_origin.clone());
+            new_provider.fee = fee.unwrap_or(old_provider.fee);
+            new_provider.payee = payee.unwrap_or(old_provider.payee);
+            new_provider.balance += self.env().transferred_value();
+            new_provider.dataset_id = dataset_id.unwrap_or(old_provider.dataset_id);
+            new_provider.dataset_id_content =
+                dataset_id_content.unwrap_or(old_provider.dataset_id_content);
+
+            // dataset content id cannot be equal to dataset id
+            if new_provider.dataset_id != default_dataset_id
+                && new_provider.dataset_id_content == new_provider.dataset_id
+            {
+                return err!(Error::DatasetIdSolutionsSame);
             }
 
-            self.check_provider_fee(fee)?;
+            // update the dataset mapping to provider
+            // remove old mapping
+            self.datasets.remove(old_provider.dataset_id);
+            if new_provider.dataset_id != default_dataset_id {
+                // insert new mapping if not the default hash, as this is used as a placeholder value
+                self.datasets
+                    .insert(new_provider.dataset_id, &provider_account);
+            }
+
+            // if the provider is
+            // not deactivating
+            // has a balance >= provider_stake_default
+            // has a dataset_id
+            // has a dataset_id_content
+            new_provider.status = if new_provider.balance >= self.provider_stake_default
+                && new_provider.dataset_id != default_dataset_id
+                && new_provider.dataset_id_content != default_dataset_id
+                && !deactivate
+            {
+                // then set the status to active
+                GovernanceStatus::Active
+            } else {
+                // else set the status to deactivated
+                GovernanceStatus::Deactivated
+            };
+
+            let old_service_origin_hash = self.hash_vec_u8(&old_provider.service_origin);
+            let new_service_origin_hash = self.hash_vec_u8(&new_provider.service_origin);
+            if old_service_origin_hash != new_service_origin_hash {
+                // updating the service origin, so check whether the new origin is available
+                if self.service_origins.contains(new_service_origin_hash) {
+                    return err!(Error::ProviderServiceOriginUsed);
+                } // else available
+            }
 
-            // provider cannot be the admin
-            self.check_not_admin(provider_account)?;
+            self.service_origins.remove(old_service_origin_hash);
```

Comment:
Shouldn't this be conditional depending on whether the service origin has changed, instead of always removing and inserting? Thinking in terms of gas fees.